cpp/src/gcylon/sorting/sorting.cpp

```diff
@@ -0,0 +1,263 @@
+    }
+
+    merged_table = cudf::merge(tables_to_merge, key_cols, column_orders);
+    if (merged_table == nullptr) {
+        return cylon::Status(cylon::Code::ExecutionError, "merge failed");
+    }
+
+    return cylon::Status::OK();
+}
+
+cylon::Status DetermineSplitPoints(const cudf::table_view &splitter_tv,
+                                   std::vector<std::unique_ptr<cudf::table>> &gathered_tables,
+                                   const std::vector<cudf::order> &column_orders,
+                                   std::shared_ptr<cylon::CylonContext> ctx,
+                                   std::unique_ptr<cudf::table> &split_points_table) {
+
+    //todo: when the number of workers is more than 10, we can sort instead of merging
+    std::unique_ptr<cudf::table> merged_table;
+    RETURN_CYLON_STATUS_IF_FAILED(
+            MergeTables(splitter_tv, gathered_tables, column_orders, merged_table));
+
+    int num_split_points = ctx->GetWorldSize() - 1;
+    auto merged_tv = merged_table->view();
+    RETURN_CYLON_STATUS_IF_FAILED(
+            gcylon::SampleTableUniform(merged_tv, num_split_points, split_points_table));
+
+    return cylon::Status::OK();
+}
+
+cylon::Status gcylon::GetSplitPoints(const cudf::table_view &sample_tv,
+                                     const int splitter,
+                                     const std::vector<cudf::order> &column_orders,
+                                     std::shared_ptr<cylon::CylonContext> ctx,
+                                     std::unique_ptr<cudf::table> &split_points_table) {
+
+    std::vector<std::unique_ptr<cudf::table>> gathered_tables;
+    bool gather_from_root = false;
+    RETURN_CYLON_STATUS_IF_FAILED(
+            gcylon::net::Gather(sample_tv, splitter, gather_from_root, ctx, gathered_tables));
+
+    if (cylon::mpi::AmIRoot(splitter, ctx)) {
+        RETURN_CYLON_STATUS_IF_FAILED(
+                DetermineSplitPoints(sample_tv, gathered_tables, column_orders, ctx, split_points_table));
+    }
+
+    cudf::table_view source_tv;
+    if (cylon::mpi::AmIRoot(splitter, ctx)) {
+        source_tv = split_points_table->view();
+    }
+
+    RETURN_CYLON_STATUS_IF_FAILED(
+            gcylon::net::Bcast(source_tv, splitter, ctx, split_points_table));
+
+    return cylon::Status::OK();
+}
+
+cylon::Status gcylon::SampleTableUniform(const cudf::table_view &tv,
+                                         const int sample_count,
+                                         std::unique_ptr<cudf::table> &sampled_table) {
+
+    if (sample_count > tv.num_rows()) {
+        return cylon::Status(cylon::Code::ExecutionError,
+                             "sample_count higher than the number of rows in the table");
+    }
+
+    float step = tv.num_rows() / (sample_count + 1.0);
+    std::vector<int32_t> sample_indices;
+    float index = step;
+    for (int i = 0; i < sample_count; ++i) {
+        sample_indices.push_back(index);
+        index += step;
+    }
```

Comment:
for smaller sample counts I think this logic might not produce enough results because of this +1. Why cant we simply keep it at `floor(num_rows/ sample_count)`?

Reply:
we want to divide the rows into n equal ranges. For this to happen, the sample_count will be n-1. 
n-1 splitters divide the rows into n ranges. The first range is from (0, step). Last range is (last splitter, last row)
here, the step size shows the range size. So, we need to add +1 to sample_count when calculating the step size.